### ISSUE TITLE
"javac returned non-zero exit code", but no errors

### ISSUE THREAD

**Issue by eed3si9n** [sbt/zinc#520]
There's been a report that the user would see

```
[error] (xxx / Compile / compileIncremental) javac returned non-zero exit code
[error] Total time: 74 s, completed Mar 22, 2018 11:23:58 AM
```

but the actual error messages from Java does not get printed out.
I think I've seen some of these myself as well.

Ref https://github.com/sbt/sbt/issues/2256
Ref https://github.com/sbt/sbt/issues/3665
Ref https://github.com/sbt/zinc/issues/376

**Comment by jvican** [comment]
That would be confirmed by https://github.com/sbt/zinc/issues/376, which I closed because I couldn't reproduce.

**Comment by retronym** [comment]
To diagnose this problem further, I suggest setting `javaHome := Some(file("<path to jdk>"))`, which will [run](https://github.com/sbt/zinc/blob/3c846ad26a33cb0cf7c8bf68d969fb33a4baa30a/internal/zinc-compile-core/src/main/scala/sbt/internal/inc/javac/JavaCompiler.scala#L39-L64) `javac` in forked mode. That might workaround some buggy error reporting code in the non-forked mode.

**Comment by eed3si9n** [comment]
I sat down with the customer today with his laptop, and was able to reproduce this. On the environment, `JAVA_HOME` was set, which was setting `javaHome` to `Some(...)`. When we `unset JAVA_HOME`, we were then able to see:

```
cannot find symbol
  symbol: method blabla
  location: variable runtime of type xxx.abc
```

In other words, the bug is likely in the error handling of forked mode.

**Comment by jvican** [comment]
I finally found the root of the error here. This happens whenever `JavaThenScala` is used. The reason why it happens is because Zinc doesn't internally do `printSummary` for Java, it only does it for Scala in the `CompilerInterface`, and therefore the messages from Javac are received but unreported.

**Comment by olafurpg** [comment]
I'm hitting on this error in building a pure Java module and lack of error messages makes it near impossible to be productive while writing Java code. A workaround is to explicitly declare `javaHome.in(Compile)` as we do in the [Scalameta build.sbt](https://github.com/scalameta/scalameta/blob/9f15793aae3cb6a00e1e2d6bcbf13e9c234ea91f/build.sbt#L470-L473)

```scala
javaHome.in(Compile) := {
  Some(file(sys.props("java.home")).getParentFile)
}
```

**Comment by SethTisue** [comment]
@retronym can you advise Olafur on this?

**Comment by jvican** [comment]
Cross-linking https://github.com/sbt/zinc/issues/415, which contains an example reproducing this error. I'm closing that ticket and keeping this one open as it's richer in details.

**Comment by wpc009** [comment]
Any workaround for this issue?

**Comment by olafurpg** [comment]
@wpc009 The workaround in https://github.com/sbt/zinc/issues/520#issuecomment-410677766 works for me under sbt version 1.1.6 but not the latest sbt 1.2.x.

**Comment by raboof** [comment]
Is this fixed by #448 or is there more work to do?

**Comment by eed3si9n** [comment]
Until we find a new failure mode other than the one fixed in #448 etc, I am going to close this.

**Comment by batterseapower** [comment]
The workaround in https://github.com/sbt/zinc/issues/520#issuecomment-410677766 didn't work for me, but setting `javaHome := None;` in the `build.sbt` did let me finally see the error output from javac.
